네비 바 커스텀

Minny27 opened the pull request:
## Motivation ⍰

- 네비 바의 아이템, 타이틀, 배경색을 커스텀하기 위함

<br>

## Key Changes 🔑

- 홈에서만 스크롤 시, 네비 바의 배경, 아이템, 타이틀 색상 변경
- navigationItem.scrollEdgeAppearance는 스크롤하지 않았을 때, navigationItem.standardAppearance는 기본 appearance
- UINavigationBarAppearance는 네비 바의 타이틀과 배경색만 바꿀 수 있음
- 그래서 ScrollView.contentOffset을 입력받아 contentOffset.y가 0을 초과하는 것을 기준으로 isSelected Output으로 내보냄
- Reactive를 확장해서 isScrolled 값에 따른 UI 변화 코드를 추가하고 해당 변수와 바인딩함

네비게이션 바가 스크롤 될 때 네비게이션 바의 왼쪽, 오른쪽 아이템과 타이틀의 색상을 바꾸는 데 어려움이 있었습니다.

스크롤 했을 때의 네비게이션 바와 타이틀 색상은 `navBarAppearance`로 설정이 가능하지만, 네비게이션 바 왼쪽, 오른쪽 아이템의 대한 색상 설정은 불가한 것으로 확인되었습니다.

그래서 두 가지 시도가 있었습니다.

### 1. `ScrollView`의 `didScroll`을 입력으로 받고 스크롤이 발생할 때 네비 바 아이템의 색상을 변경해주기

- 이렇게 하면 디폴트와 스크롤했을 때의 분기를 나눌 수 없었음.

### 2. `ScrollVied`의 `contentOffset`을 입력으로 받아서 `offset`의 y가 0일 때와 0보다 클 때를 나눠서 분기하기

- 이렇게 하면 원하는 방식대로 구현할 수 있었음.
- 네비게이션 바와 타이틀 색상은 navBarAppearance로 설정
- 여기서 `scrollEdgeAppearance`는 스크롤 안 됐을 때의 `navBarAppearance`이고, `standardAppearance가` 스크롤 했을 때의 디폴트 `navBarAppearance`

코드는 다음과 같음
```swift
final class HomeViewController: NagazaBaseViewController {

    { ... }
    
    override func navigationSetting() {
        super.navigationSetting()
        let navBarAppearance = UINavigationBarAppearance()
        navBarAppearance.backgroundColor = NagazaAsset.Colors.black.color
        navBarAppearance.titleTextAttributes = [.font: UIFont.ngaH3M, .foregroundColor: NagazaAsset.Colors.white.color]
        navigationItem.scrollEdgeAppearance = navBarAppearance // 스크롤 안 했을 때 네비 바 외관 설정
        
        let mapButtonItem = UIBarButtonItem(image: NagazaAsset.Images.map.image, style: .plain, target: nil, action: nil)
        let searchButtonItem = UIBarButtonItem(image: NagazaAsset.Images.search.image, style: .plain, target: nil, action: nil)
        
        navigationItem.title = "홈"
        navigationItem.leftBarButtonItem = mapButtonItem
        navigationItem.rightBarButtonItem = searchButtonItem
    }
    
    override func bindViewModel() {
        let contentOffset = scrollView.rx.contentOffset.asDriver()
        
        let input = HomeViewModel.Input(contentOffset: contentOffset)
        
        let output = viewModel.transform(input: input)
        
        output.isScrolled
            .drive(self.rx.isScrolled)
            .disposed(by: disposeBag)
    }
}

// isScrolled라는 Driver Bool 값에 따라 UI가 변경되는 요소는 Reactive를 확장해서 바인딩되도록 구현
extension Reactive where Base: HomeViewController {
    var isScrolled: Binder<Bool> {
        return Binder(self.base) { base, value in
            if value {
                base.navigationItem.leftBarButtonItem?.tintColor = NagazaAsset.Colors.selected.color
                base.navigationItem.rightBarButtonItem?.tintColor = NagazaAsset.Colors.selected.color
            } else {
                base.navigationItem.leftBarButtonItem?.tintColor = NagazaAsset.Colors.white.color
                base.navigationItem.rightBarButtonItem?.tintColor = NagazaAsset.Colors.white.color
            }
        }
    }
}

// HomeViewModel
final class HomeViewModel: ViewModelType {
    struct Input {
        let contentOffset: Driver<CGPoint>
    }
    
    struct Output {
        let isScrolled: Driver<Bool>
    }
    
    // contentOffset을 input으로 받아서 contentOffset.y 가 0초과 조건을 기준으로 Bool로 리턴
    func transform(input: Input) -> Output {
        let isScrolled = input.contentOffset
            .map { point in
                return point.y > 0
            }
        
        return Output(isScrolled: isScrolled)
    }
}
```

미리보기는 다음과 같음

<img src="https://github.com/NAGAZA-Team/NAGAZA-iOS/assets/68800789/67a40f15-14e0-491a-9247-50c10e78bc31" width=25%>



더 좋은 아이디어 있으시면 의견주시거나 commit 올리셔도 됩니다!

<br>

## To Reviewers 🙏🏻

- [x] 홈에서 정상적으로 스크롤 시, 타이틀, 배경, 아이템 색상 변경되는 지 봐주세요.

Minny27 committed:
Add: 색상 및 이미지 리소스 추가

Minny27 committed:
Feat: 탭 바 커스텀

- 홈에서만 스크롤 시, 네비 바의 배경, 아이템, 타이틀 색상 변경
- navigationItem.scrollEdgeAppearance는 스크롤하지 않았을 때, navigationItem.standardAppearance는 기본 appearance
- UINavigationBarAppearance는 네비 바의 타이틀과 배경색만 바꿀 수 있음
- 그래서 ScrollView.contentOffset을 입력받아 contentOffset.y가 0을 초과하는 것을 기준으로 isSelected Output으로 내보냄
- Reactive를 확장해서 isScrolled 값에 따른 UI 변화 코드를 추가하고 해당 변수와 바인딩함

cestbonciel reviewed (approved):
고생하셨습니다! PR 하신 부분 차후 머지하시면 pull 해서, 홈 부분 스크롤시 아이템 색상 변경에 대해서 더 살펴보도록 할게요.

Jeon0976 committed:
Feat: 그라디언드 기능 적용

Jeon0976 commented:
그라디언트 기능 적용 commit 했습니다 확인 부탁드리겠습니다
![Simulator Screen Recording - iPhone 14 Pro Max - 2023-12-12 at 15 03 27](https://github.com/NAGAZA-Team/NAGAZA-iOS/assets/113761591/be2a2cc9-92f3-4684-a923-ffc11afe0d9d)

Jeon0976 reviewed (approved):
그라디언트 기능 추가 했습니다

hililyy committed:
Update: Gradient 디자인 수정

Jeon0976 committed:
Refector: Enum 기준에서 Alpha 기준으로 navi appearance 변경 되도록 수정

Jeon0976 commented:
변경 후 화면입니다.
![Uploading Simulator Screen Recording - iPhone SE (3rd generation) - 2023-12-13 at 08.32.50.gif…]()

Jeon0976 committed:
Refector: alpha값을 정하는 contentOffSet 최소 기준 70 -> 0으로 변경

Minny27 committed:
Enhance: 스크롤 contentOffset의 y 값에 따른 enum으로 분리

- 프레임 높이의 1/3이전과 이후로 나눔
- 프레임 높이의 1/3지점 까지만 alpha 적용
- 추천 테마 높이 프레임 높이 1/2으로 설정
- NagazaSiza에서 CGFloat 확장으로 변경
- window에서 프레임 높이, 너비 확장
- 불필요한 코드 삭제

Minny27 commented:
추천 테마 뷰 높이를 디바이스 높이 기준 1/2로 설정하고
스크롤 알파값은 디바이스 높이의 1/3 지점까지로 설정해서 수정했습니다
